utils/generate_protocol_types.py

```diff
@@ -0,0 +1,450 @@
+        :return: line with resolved forward reference
+        """
+        return re.sub(self._forward_ref_re, partial(self._resolve_forward_ref_re_sub_repl, fw_ref=fw_ref), line)
+
+    async def _retrieve_top_level_domain(self):
+        browser = await launch(args=['--no-sandbox', '--disable-setuid-sandbox'])
+        base_endpoint = re.search(r'ws://([0-9A-Za-z:.]*)/', browser.wsEndpoint).group(1)
+        page = await browser.newPage()
+
+        logger.info(f'Loading raw protocol specification')
+        t_start = time.perf_counter()
+
+        await page.goto(f'http://{base_endpoint}/json/protocol')
+        page_content = await page.evaluate('document.documentElement.innerText')
+        try:
+            await browser.close()
+        except Exception as e:
+            logger.warning(f'Exception on browser close: {e}')
+
+        logger.info(f'Loading raw protocol specification in {time.perf_counter()-t_start:.2f}s')
+        self.domain = json.loads(page_content)
+
+    def retrieve_top_level_domain(self):
+        """
+        Fetches as sets the class variable domains for later use.
+        :return: None
+        """
+        asyncio.get_event_loop().run_until_complete(self._retrieve_top_level_domain())
+
+    def gen_spec(self):
+        """
+        Generate the Protocol class file lines within self.code_gen attribute. Uses an IndentManager context manager to 
+        keep track of the current indentation level. Resolves all forward references. Expands self-recursive types to
+        RECURSIVE_TYPE_EXPANSION_DEPTH. Expands cyclic types to 1 level.
+        :return: None
+        """
+
+        self.code_gen.insert_before_code('"""')
+        self.code_gen.insert_before_code(self.header)
+        self.code_gen.insert_before_code('"""')
+        self.code_gen.add_newlines(num=2)
+
+        logger.info(f'Generating protocol spec')
+        t_start = time.perf_counter()
+
+        self.code_gen.add('class Protocol:')
+        with self.code_gen.indent_manager:
+            for domain in self.domain['domains']:
+                domain_name = domain['domain']
+                self.code_gen.add(f'class {domain_name}:')
+                self.code_gen.add_comment_from_info(domain)
+                self.all_known_types[domain_name] = domain_known_types = {}
+                with self.code_gen.indent_manager:
+                    for type_info in domain.get('types', []):
+                        self.code_gen.add_comment_from_info(type_info)
+                        item_name = type_info["id"]
+                        if 'properties' in type_info:
+                            # name mangled to avoid collisions
+                            # _type = f'_{domain_name}_{type_info["id"]}'
+                            td = self.generate_typed_dicts(type_info, domain_name)
+                            self.typed_dicts.update(td)
+                            _type = [*td.keys()][0]
```

Comment:
nit pick: for reserved keywords python suggests using trailing underscore rather than leading one. i.e. `type_` instead of `_type` :)